### PR TITLE
Extracting external sites

### PR DESCRIPTION
**Pull request by bmenn**
We need functionality to identify other potential sites to scrape.

### CODE REVIEW

**Comment by bmenn** [comment]
If someone get to this before I get back to this, we could use SQLAlchemy data model objects to simplify data access (but this does not mean less code).

http://docs.sqlalchemy.org/en/latest/orm/extensions/automap.html?highlight=automap#module-sqlalchemy.ext.automap

**Comment by bmenn** [comment]
@dlrobertson @egrossman 

Not 100% done here, but almost. Need to add a simple integration test and I'll be good. There's also an example of unit testing a Luigi task as well.

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
```
Comment (dlrobertson) [inline Vagrantfile:2]: :+1:

File: Vagrantfile [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  # information on available options.
+
+  # Enable provisioning with a shell script. Additional provisioners such as
+  # Puppet, Chef, Ansible, Salt, and Docker are also available. Please see the
+  # documentation for more information about their specific syntax and use.
+  config.vm.provision "shell", inline: <<-SHELL
+    apt-get update
+    apt-get install -y postgresql postgresql-contrib
+    echo "listen_addresses = '*'" | sudo tee -a /etc/postgresql/9.5/main/postgresql.conf
+    echo "host all postgres 0.0.0.0/0 trust" | sudo tee /etc/postgresql/9.5/main/pg_hba.conf
```
Comment (dlrobertson) [inline Vagrantfile]: Do we really need it this open? I'd think `localhost` would be good enough for both listen and no password auth.

File: htetl/tasks/sites.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        out = []
+        for df_index, row in pages.iterrows():
+            site_list = sites.extract_sites(row['Content'])
+            out.extend([
+                {
+                'PageId': row.id,
+                'ExternalSiteId': base_sites.iloc[external_site]['Id']
+                }
+                for external_site in site_list
```
Comment (dlrobertson) [inline htetl/tasks/sites.py:48]: :boom:

File: tests/integration/__init__.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import luigi
+import os
+import pandas as pd
+import sqlalchemy
+import sys
+if sys.version_info[0] < 3:
+    from StringIO import StringIO
+else:
+    from io import StringIO
```
Comment (dlrobertson) [inline tests/integration/__init__.py]: Is this the more pythonic way? I thought the standard way of doing this was

```python
try:
    import x
except ImportError:
    import y
````

File: tests/integration/__init__.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        )
+    )
+
+    test_config = luigi.configuration.LuigiConfigParser.instance()
+    test_config.set('QueryPostgres', 'host', POSTGRES_HOST)
+    test_config.set('QueryPostgres', 'database', POSTGRES_DATABASE)
+    test_config.set('QueryPostgres', 'user', POSTGRES_USER)
+    test_config.set('QueryPostgres', 'password', POSTGRES_PASSWORD)
+
+    # meta = sqlalchemy.MetaData(engine)
```
Comment (dlrobertson) [inline tests/integration/__init__.py]: Is this going to be used. If not I'm a fan of just removing it

File: tests/integration/__init__.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    pd.read_csv(
+        StringIO(test_data.RAWPAGEDATA_RAW__PAGE_CSV_CONTENTS)
+    ).to_sql('Page', engine)
+    pd.read_csv(
+        StringIO(test_data.BASESITES_BASE_SITES__CSV_CONTENTS)
+    ).to_sql('BaseSites', engine)
+#
+#
+# def teardown_package():
```
Comment (dlrobertson) [inline tests/integration/__init__.py]: Same

File: tests/unit/tasks/mocks.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import htetl.tasks.sites as tasks_sites
+from htetl.tasks.loadpages import RawPageData
+
+
+def setup_RawPageData():
```
Comment (dlrobertson) [inline tests/unit/tasks/mocks.py]: Isn't a `pass` needed here?

File: tests/integration/__init__.py (hunk 4 above)
Comment (bmenn) [reply tests/integration/__init__.py]: Yes, that is more Pythonic. Stole the line from SO without thinking about that.

File: tests/integration/__init__.py (hunk 5 above)
Comment (bmenn) [reply tests/integration/__init__.py]: Should be gone now.

File: Vagrantfile (hunk 2 above)
Comment (bmenn) [reply Vagrantfile]: We need this because localhost isn't the dev's localhost when using Vagrant.

File: tests/unit/tasks/mocks.py (hunk 7 above)
Comment (bmenn) [reply tests/unit/tasks/mocks.py]: Honestly, I don't know why this file is here.

File: tests/unit/tasks/test_sites.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+from ... import test_data, test_utils
+
+
+def setup_module():
+    test_config = luigi.configuration.LuigiConfigParser.instance()
+    test_config.set('QueryPostgres', 'host', 'fake_host')
+    test_config.set('QueryPostgres', 'database', 'fake_db')
+    test_config.set('QueryPostgres', 'user', 'fake_user')
+    test_config.set('QueryPostgres', 'password', 'fake_password')
+
```
Comment (egrossman) [inline tests/unit/tasks/test_sites.py:15]: Is setup_module used anywhere or is this ran automatically by luigi?

**Review by egrossman — APPROVED** [review]
It took me way too long to figure out how mock testing works, but finally got through the code. I noticed that extacts.sites isn't a luigi task, but just methods to be called in tasks. Is that how you want us to refactor our code too?

File: tests/unit/tasks/test_sites.py (hunk 8 above)
Comment (bmenn) [reply tests/unit/tasks/test_sites.py:15]: `setup_module` is used by nose. A little more details here if you are interested http://nose.readthedocs.io/en/latest/writing_tests.html.

**Pull request merged**